crates/solver/src/solver/http_solver/instance_cache.rs

```diff
@@ -0,0 +1,187 @@
+use super::{instance_creation::InstanceCreator, settlement::SettlementContext};
+use crate::{s3_instance_upload::S3InstanceUploader, solver::Auction};
+use model::auction::AuctionId;
+use shared::http_solver::model::BatchAuctionModel;
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+/// To `Driver` every http solver is presented as an individual `Solver` implementor. Internally
+/// http solvers share the same data that is needed to create the instance for the same auction. In
+/// order to waste less resources we create the instance once per auction in this component.
+pub struct SharedInstanceCreator {
+    creator: InstanceCreator,
+    uploader: Option<Arc<S3InstanceUploader>>,
+    last: Mutex<Option<Cache>>,
```

Comment:
I don't expect it to matter performance wise but I think a `RwLock` might be a bit more appropriate here. There will only every be a single writer but possibly multiple parallel readers.

Reply:
There is going to be ~10 solvers requesting the instance. The first one locks the mutex and performs the instance creation which takes ~5 seconds. Then all of them read a copy of the instance. Copying the instance is not an expensive operation (especially with the Arc change below) so I'd expect the difference between Mutex and RwLock to be unnoticeable.

Reply:
I think a `RwLock` isn't a super obvious gain to me here, because all 10 solvers can potentially are contending for upgrading their read lock to a write lock for the initial update.

If there was a single writer, then I agree a `RwLock` would make more sense.

Reply:
Oh, yeah `RwLock` doesn't make much sense indeed. A solver can't know whether a read locks is sufficient so they would all have to request a write lock anyway. But as Valentin mentioned with the `Arc` change contention should be minimal anyway.